infrastructures/docker/environment/src/main/java/org/eclipse/che/workspace/infrastructure/docker/environment/compose/ComposeEnvironmentFactory.java

```diff
@@ -116,14 +121,32 @@ void addRamLimitAttribute(
         machineConfig = new InternalMachineConfig();
         machines.put(entry.getKey(), machineConfig);
       }
-      final Map<String, String> attributes = machineConfig.getAttributes();
-      if (isNullOrEmpty(attributes.get(MEMORY_LIMIT_ATTRIBUTE))) {
-        final Long ramLimit = entry.getValue().getMemLimit();
-        if (ramLimit != null && ramLimit > 0) {
-          attributes.put(MEMORY_LIMIT_ATTRIBUTE, String.valueOf(ramLimit));
-        } else {
-          attributes.put(MEMORY_LIMIT_ATTRIBUTE, defaultMachineMemorySizeAttribute);
-        }
+
+      initIfEmpty(
+          () -> entry.getValue().getMemLimit(),
+          machineConfig,
+          MEMORY_LIMIT_ATTRIBUTE,
+          defaultMachineMaxMemorySizeAttribute);
+      initIfEmpty(
+          () -> entry.getValue().getMemRequest(),
+          machineConfig,
+          MEMORY_REQUEST_ATTRIBUTE,
+          defaultMachineRequestMemorySizeAttribute);
+    }
+  }
+
+  private void initIfEmpty(
```

Comment:
In a case when the memory limit is set in the recipe does this code respects it?